### ISSUE TITLE
[Question] How to implement multiple stores with redux-toolkit?

### ISSUE THREAD

**Issue by SurjitSahoo** [reduxjs/redux-toolkit#3000]
I'm working on a micro-front-ends architecture app, where one shell app will render multiple remote apps under single roof.
I want to have an mf as a centralized redux store to which all the remote apps can connect to share common state.
And remote apps should also have their own redux stores for their local states...

Here's the issue..
if central store and local store both have a state called `count`, updating this state in any of the stores, updates both of them, they behave as if it's the exact same state.

I tried changing the store names.. like `centralStore` and `localStore`, it didn't have any effect.
I tried changing the slice name.. then the nearest store worked and the other store stopped working..

in the following example, localStore works.. and the centralStore stops working.
```jsx
function CentralStoreProvider({children}) {
    return <Provider store={centralStore}>{children}<Provider />
}

// app.js
<CentralStoreProvider>
    <Provider store={localStore}>
        <App />
    <Provider />
<CentralStoreProvider>
```

**Comment by markerikson** [comment]
React-Redux normally expects that there's only _one_ Redux store, and passes it down by context.  By default components can _only_ read that one store.  So, no, you can't have multiple stores being read in components like that.

**Comment by SurjitSahoo** [comment]
> React-Redux normally expects that there's only _one_ Redux store, and passes it down by context. By default components can _only_ read that one store. So, no, you can't have multiple stores being read in components like that.

But redux allows to have multiple stores. So is it react-redux specific?

**Comment by markerikson** [comment]
@SurjitSahoo it's technically possible to create multiple Redux stores in an app, yes, but we have _always_ told users that there _should_ only be _one_ Redux store:

- https://redux.js.org/style-guide/#only-one-redux-store-per-app

**Comment by blwinters** [comment]
If I understand your use case correctly, why not just use separate slices instead of separate stores? You can use RTKQuery to setup a separate api slice for each app / baseUrl: 
https://stackoverflow.com/questions/68152504/should-i-use-rtk-query-per-base-url-or-resource

**Comment by phryneas** [comment]
Generally, you *can* have multiple stores with different contexts and different hooks. This is described in [custom context](https://react-redux.js.org/api/hooks#custom-context). But if you can avoid it, avoid id.

**Comment by SurjitSahoo** [comment]
> Generally, you _can_ have multiple stores with different contexts and different hooks. This is described in [custom context](https://react-redux.js.org/api/hooks#custom-context). But if you can avoid it, avoid id.

Thanks.. It solved my issue..

Basically, I want to have a way to share some common states across mf apps, without having any side-efect on those apps, if they need something, they can just import it from central-store.. they don't even need to know is it redux or something else.

Thanks again …

**Comment by johnzenith** [comment]
It is possible to have one store shared store across your mono-repo and isolated store each for app.

Things to note:

- You have to use one `Provider`
- Use the `configureStore` as usual to setup the shared store without needing to put it in a `Provider`
- Export the shared store and reducers
- In each app, you need to import the store and reducers, and merge it. By merging, you just need to spread the `shared store` object and merge it with the app store: `const store = {...sharedStore, ...appStore}`

Note: Redux uses a single object instance, so even though configureStore will only create a single instance.

`js

import { sharedStore, sharedReducers } from '/mono-repo/path'

export const store = configureSotre({
   reducer: {
       ...sharedReducers,
      appReducer1,
      appReduer2,
     ...
   }
})
`

Note: ensure your shared store reducers are unique and do not conflict with individual app reducers.

**Comment by EskiMojo14** [comment]
@johnzenith spreading stores into each other makes no sense. all of the methods (getStore, dispatch, etc) on sharedStore will be overwritten by the methods from appStore.

**Comment by johnzenith** [comment]
> @johnzenith spreading stores into each other makes no sense. all of the methods (getStore, dispatch, etc) on sharedStore will be overwritten by the methods from appStore.

They won't be, like I mentioned, unless the reducers are the same.

**Comment by markerikson** [comment]
@johnzenith : not the reducers, the _methods_:

```ts
const mergedStore = {...store1, ...store2}
console.log(mergedStore.dispatch === store2.dispatch) // true
console.log(mergedStore.dispatch === store1.dispatch) // false
```

Don't do that!

**Comment by johnzenith** [comment]
> [@johnzenith](https://github.com/johnzenith) : not the reducers, the _methods_:
> 
> const mergedStore = {...store1, ...store2}
> console.log(mergedStore.dispatch === store2.dispatch) // true
> console.log(mergedStore.dispatch === store1.dispatch) // false
> Don't do that!

Yes, you're right. The reducers needs to be spread.

**Comment by johnzenith** [comment]
Hey @markerikson - what do you think about spreading the reducers?

**Comment by johnzenith** [comment]
I believe you mean this:

```js
function overrideStoreMethodsDirectly(storeToOverride: Store, storeToOverrideWith: Store): void {
  storeToOverride.dispatch = storeToOverrideWith.dispatch.bind(storeToOverrideWith);
  storeToOverride.getState = storeToOverrideWith.getState.bind(storeToOverrideWith);
  storeToOverride.subscribe = storeToOverrideWith.subscribe.bind(storeToOverrideWith);
  storeToOverride.replaceReducer = storeToOverrideWith.replaceReducer.bind(storeToOverrideWith);
}
```

**Comment by markerikson** [comment]
@johnzenith I... am very lost at what you're trying to accomplish at this point.  Why are you doing that?  What's the purpose?

**Comment by EskiMojo14** [comment]
> Hey [@markerikson](https://github.com/markerikson) - what do you think about spreading the reducers?

spreading common reducers is a valid way of using the same structure in multiple stores, they will not have the same state however.

> I believe you mean this:
> 
> function overrideStoreMethodsDirectly(storeToOverride: Store, storeToOverrideWith: Store): void {
>   storeToOverride.dispatch = storeToOverrideWith.dispatch.bind(storeToOverrideWith);
>   storeToOverride.getState = storeToOverrideWith.getState.bind(storeToOverrideWith);
>   storeToOverride.subscribe = storeToOverrideWith.subscribe.bind(storeToOverrideWith);
>   storeToOverride.replaceReducer = storeToOverrideWith.replaceReducer.bind(storeToOverrideWith);
> }

this is nonsensical. redux uses closure factories not classes so `.bind` is irrelevant, and all you're doing here is mutating one store with the methods of another.

**Comment by johnzenith** [comment]
Oh, sorry. I pasted the wrong code test. I have updated the structure.

I wanted to see how the shared store state sync up across a mono-repo app. So since each individual is receiving the reducers from the shared store states, so I want to know whether mutating the shared store state object to keep it in sync makes sense.

The why I'm mutating the shared store to reference the app store is that on the app itself, mutations will be on the app store object, and doing `sharedStore.getState()` won't reflect the current state in most cases.
